<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-rafaelsangarcia/blob/main/Actividad3_A01182780.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#ACTIVIDAD 3
**RAFAEL SANCHEZ GARCIA**


#PART 1

**Fundamentos de bases de datos y ciencia de datos**
200 palabras

#PART 2

## LOAD DATA

In [39]:
#LOAD THE DATABASE TO MY COLAB
input = "https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv"

In [210]:
import pandas as pd
import numpy as np
df = pd.read_csv(input)
df.head()

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

**Lets verify is we have NAs**

In [41]:
df.isnull().values.any()

True

**We have Null values, so we need to start to decide what to do which them. But first lets explore which columns have nulls**

In [42]:
df.isnull().any()

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

**All except ID and X1 has nulls**

## CLEAN DATA

**Lets analysis the data**


In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

**ID and X1 are int, the rest are float,Lets see how are the distribution**

In [44]:
#better way to see the data
df.describe().transpose()

count           mean            std       min       25%       50%  \
ID   30000.0   15000.500000    8660.398374       1.0   7500.75   15000.5   
X1   30000.0  167484.322667  129747.661567   10000.0  50000.00  140000.0   
X2   29999.0       1.603753       0.489125       1.0      1.00       2.0   
X3   29998.0       1.853057       0.790320       0.0      1.00       2.0   
X4   29998.0       1.551903       0.521968       0.0      1.00       2.0   
X5   29995.0      35.484214       9.218024      21.0     28.00      34.0   
X6   29997.0      -0.016635       1.123829      -2.0     -1.00       0.0   
X7   29995.0      -0.133689       1.197254      -2.0     -1.00       0.0   
X8   29993.0      -0.166405       1.196048      -2.0     -1.00       0.0   
X9   29991.0      -0.220800       1.169153      -2.0     -1.00       0.0   
X10  29984.0      -0.266342       1.133296      -2.0     -1.00       0.0   
X11  29986.0      -0.291136       1.150134      -2.0     -1.00       0.0   
X12  29989.0   51236.862750   73645.219278 -165580.0   3565.00   22387.0   
X13  29989.0   49190.734669   71183.385123  -69777.0   2986.00   21207.0   
X14  29987.0   47025.350152   69360.863317 -157264.0   2667.50   20089.0   
X15  29985.0   43275.652326   64345.500073 -170000.0   2329.00   19052.0   
X16  29983.0   40324.493980   60809.984983  -81334.0   1763.50   18107.0   
X17  29990.0   38881.135745   59561.312967 -339603.0   1256.25   17081.0   
X18  29992.0    5662.945886   16564.165089       0.0   1000.00    2100.0   
X19  29991.0    5922.488913   23044.177075       0.0    835.50    2009.0   
X20  29992.0    5225.623400   17608.422625       0.0    390.00    1800.0   
X21  29989.0    4827.252526   15668.751975       0.0    296.00    1500.0   
X22  29989.0    4800.297209   15280.842069       0.0    251.00    1500.0   
X23  29995.0    5216.259977   17778.848359       0.0    118.00    1500.0   
Y    29997.0       0.221189       0.415054       0.0      0.00       0.0   

           75%        max  
ID    22500.25    30000.0  
X1   240000.00  1000000.0  
X2        2.00        2.0  
X3        2.00        6.0  
X4        2.00        3.0  
X5       41.00       79.0  
X6        0.00        8.0  
X7        0.00        8.0  
X8        0.00        8.0  
X9        0.00        8.0  
X10       0.00        8.0  
X11       0.00        8.0  
X12   67139.00   964511.0  
X13   64027.00   983931.0  
X14   60182.00  1664089.0  
X15   54560.00   891586.0  
X16   50213.00   927171.0  
X17   49208.25   961664.0  
X18    5006.00   873552.0  
X19    5000.00  1684259.0  
X20    4505.50   896040.0  
X21    4014.00   621000.0  
X22    4033.00   426529.0  
X23    4000.00   528666.0  
Y         0.00        1.0

### **Lets start with column Y**

In [282]:
ndf = df.copy()

In [283]:
ndf['Y'].isnull().any()

True

In [284]:
ndf[ndf['Y'].isnull()]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
6232    6233  60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...  NaN  NaN   
29824  29825  40000  1.0  1.0  1.0  47.0  2.0  2.0  2.0  2.0  ...  NaN  NaN   
29825  29826  50000  1.0  2.0  1.0  41.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   

       X17  X18  X19  X20  X21  X22  X23   Y  
6232   NaN  NaN  NaN  NaN  NaN  NaN  NaN NaN  
29824  NaN  NaN  NaN  NaN  NaN  NaN  NaN NaN  
29825  NaN  NaN  NaN  NaN  NaN  NaN  NaN NaN  

[3 rows x 25 columns]

We can see that the item 6232, 29824 and 29825 has a lots of NaN, we can eliminate, and it will not affect too much to eliminate because it is just 3 items out 30,000

In [285]:
ndfIndex = ndf[ndf['Y'].isnull()].index
ndfIndex
#ndf = ndf.drop(6232)
ndf =ndf.drop(ndfIndex);

In [286]:
ndf[ndf['Y'].isnull()]

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

In [287]:
ndf['Y'].isnull().any()

False

**Column Y does not have more null values.**